js/packages/mobile-wallet-adapter-protocol/android/src/main/java/com/solanamobile/mobilewalletadapter/reactnative/SolanaMobileWalletAdapterModule.kt

```diff
@@ -0,0 +1,210 @@
+import com.solana.mobilewalletadapter.clientlib.scenario.LocalAssociationScenario
+import com.solana.mobilewalletadapter.clientlib.scenario.Scenario
+import kotlinx.coroutines.*
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.Semaphore
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import java.util.concurrent.ExecutionException
+
+class SolanaMobileWalletAdapterModule(reactContext: ReactApplicationContext) :
+    ReactContextBaseJavaModule(reactContext), CoroutineScope {
+
+    override val coroutineContext =
+        Dispatchers.IO + CoroutineName("SolanaMobileWalletAdapterModuleScope") + SupervisorJob()
+
+    companion object {
+        private const val ASSOCIATION_TIMEOUT_MS = 10000
+        private const val CLIENT_TIMEOUT_MS = 90000
+        private var localAssociation: LocalAssociationScenario? = null
+        private var mobileWalletAdapterClient: MobileWalletAdapterClient? = null
+        private var semTerminated: Semaphore? = null
+        private val mutex = Mutex()
+    }
+
+    override fun getName(): String {
+        return "SolanaMobileWalletAdapter"
+    }
+
+    @Throws(JSONException::class)
+    private fun convertMapToJson(readableMap: ReadableMap?): JSONObject {
+        val json = JSONObject()
+        val iterator = readableMap!!.keySetIterator()
+        while (iterator.hasNextKey()) {
+            val key = iterator.nextKey()
+            when (readableMap.getType(key)) {
+                ReadableType.Null -> json.put(key, JSONObject.NULL)
+                ReadableType.Boolean -> json.put(key, readableMap.getBoolean(key))
+                ReadableType.Number -> json.put(key, readableMap.getDouble(key))
+                ReadableType.String -> json.put(key, readableMap.getString(key))
+                ReadableType.Map -> json.put(key, convertMapToJson(readableMap.getMap(key)))
+                ReadableType.Array -> json.put(
+                    key,
+                    readableMap.getArray(key)?.let { convertArrayToJson(it) }
+                )
+            }
+        }
+        return json
+    }
+
+    @Throws(JSONException::class)
+    private fun convertArrayToJson(readableArray: ReadableArray): JSONArray {
+        val array = JSONArray()
+        for (i in 0 until readableArray.size()) {
+            when (readableArray.getType(i)) {
+                ReadableType.Null -> {}
+                ReadableType.Boolean -> array.put(readableArray.getBoolean(i))
+                ReadableType.Number -> array.put(readableArray.getDouble(i))
+                ReadableType.String -> array.put(readableArray.getString(i))
+                ReadableType.Map -> array.put(convertMapToJson(readableArray.getMap(i)))
+                ReadableType.Array -> array.put(convertArrayToJson(readableArray.getArray(i)))
+            }
+        }
+        return array
+    }
+
+    @Throws(JSONException::class)
+    private fun convertJsonToMap(jsonObject: JSONObject): WritableMap {
```

Comment:
Could the return of this method be an immutable type (like `Map`)?
(Except if the mutability is used upstream)

Reply:
Return type: done.

The choice of type is not mine. This is the type that React Native demands when you're passing object/array data across the bridge to JS: https://reactnative.dev/docs/native-modules-android#argument-types